modules/git/repo_tag.go

```diff
@@ -35,34 +36,66 @@ func (repo *Repository) CreateTag(name, revision string) error {
 	return err
 }
 
+// CreateAnnotatedTag create one annotated tag in the repository
+func (repo *Repository) CreateAnnotatedTag(name, message, revision string) error {
+	_, err := NewCommand("tag", "-a", "-m", message, name, revision).RunInDir(repo.Path)
+	return err
+}
+
 func (repo *Repository) getTag(id SHA1) (*Tag, error) {
 	t, ok := repo.tagCache.Get(id.String())
 	if ok {
 		log("Hit cache: %s", id)
-		return t.(*Tag), nil
+		tagClone := *t.(*Tag)
+		return &tagClone, nil
 	}
 
-	// Get tag type
-	tp, err := NewCommand("cat-file", "-t", id.String()).RunInDir(repo.Path)
+	// Get tag name
+	name, err := repo.GetTagNameBySHA(id.String())
 	if err != nil {
 		return nil, err
 	}
-	tp = strings.TrimSpace(tp)
 
-	// Tag is a commit.
+	tp, err := repo.GetTagType(id)
+	if err != nil {
+		return nil, err
+	}
+
+	// The id passed in could be a tag object ID or a commit ID, so getting both
+	tagID := id
+	if tagIDStr, _ := repo.GetTagID(name); tagIDStr != "" {
```

Comment:
I don't quite like that this ignores all errors, it should probably check for not exists error, same for other places below

Reply:
Ok, yeah, actually wanted to revisit that and make it more clear and handle the errors. Now it should be standard.